### PR TITLE
A cleaning step in using heuristics for inference of the return clause

### PR DESCRIPTION
**Pull request by herbelin**
This started when Jason noticed on coq-club that "refine (fun H1 => match H1 with eq_refl => _ end)" was not using "small inversion" while "intro H1; refine (match H1 with eq_refl => _ end)" was.

I then took time to clean the organization of the three heuristics used for inference of the return clause, resulting in the attached 4 commits, progressively unifying the use of heuristics, depending on whether there is a type constraint or not, on whether this type constraint is referring to variables by rels or vars, trying the heuristics in the same order in each situation.

Details are in the commit logs.

It seems ok to me for 8.6 in the sense that it is cleaning an existing framework. No effect on the contribs tested by the jenkins system even though there is a small semantic change: when there is a type constraint with contains "rel" variables, priority was given to abstracting this type constraint over these "rel" variables without inverting even if using a "small inversion" was relevant (as in Jason's example). Now, "small inversion" is always tried first. 8.7 could be considered too if needed, though I would advocate for a quick integration to limit merging issues.

### CODE REVIEW

File: pretyping/cases.ml
Comment (mattam82) [inline pretyping/cases.ml]: Shouldn't we have isRel or isVar here then ?

**Comment by mattam82** [comment]
I like it very much! This makes it much clearer which strategies are applied and is more uniform. From what I understand, the dependency strategies are still overlapping, but this depends only on the type of the terms to match right? So we could specify that if there are indices with not enough pattern structure we only try 2 and 3 otherwise we stick to 1 and 3?

**Comment by silene** [comment]
(This is not directly related to the patch, but the patch makes it more apparent, at least to me.) Am I right that several strategies are precomputed but only the first successful one is kept? Then wouldn't it make to sense to use a lazy list here, so that strategies are computed only when all the previous ones have failed?

**Comment by herbelin** [comment]
@mattam: We might need more time for a thorough discussion, but to be short, somehow, is your question the question of what heuristic to prefer when several apply, and how to specify it, if ever we want to specify something? For definitions with a type constraint, this does not matter since what matters is the existence of a predicate which satisfies the external constraint and the inner constraints in the branch, whatever the predicate is. For refining, this matters and we fall again on the discussion about unification, whether it should make choices or leave open constraints. For instance, in the UIP issue, instead of using one or the other strategy, one could expect that

`Goal forall (T : Type) (feq : Eq=Eq->T) (H1 : Eq=Eq), feq eq_refl = feq H1.`
`intros.`
`refine (match H1 with eq_refl => 0 end).`

returns goal

`============================`
`feq eq_refl = feq ?p@{H1:=eq_refl}`

with unification constraint

`?p@{H1:=H1} == H1`

and that the constraint is solved only at the time of the forthcoming `reflexivity`.

So, there is a tension here between flexibility (i.e. keeping unification constraint) and simplicity (taking decisions which we hope to be the most natural ones but with the risk of taking the bad decision). Here (1) intrinsically follows the "flexible" approach and, the choice which is made, if ever one is made, it is the unification algorithm which makes it, while (2) and (3) take decisions in advance, so that the unification algorithm is not involved anymore. 

@silene: Strictly speaking, we indeed compute in advance strategies which may be not tried. I feel myself lazy in trying to gain some cycles here, even more than I'm unsure this is critical time, but I agree that it could be made more optimized. (In particular I'm supporting you if you want to try something.)

**Comment by herbelin** [comment]
Do we want more discussion on this? It is somehow more a cleaning than a fix strictly speaking. Do we want it for 8.6 or for 8.7?

**Comment by herbelin** [comment]
No comments on rather merging to 8.6, so I'm considering favoring merging to 8.6 based on that it had no effect on the bench of contribs, and that this would limit merging issues, and that this would be made available quicklier.

**Comment by mattam82** [comment]
Apart from the possibly missing isVar I mentioned I don't see any issue here. @JasonGross could you try it on fiat to see if it has any adverse effect there before we merge, just to be safe?

**Comment by JasonGross** [comment]
First I need to get fiat to build with 8.6 again.

However, [fiat-crypto](https://github.com/mit-plv/fiat-crypto) should build with Coq 8.6 right now.  Is there a way to get this to be part of the automated test-suite?

**Comment by mattam82** [comment]
Ah ok, I thought that was what Paul used. Certainly, I guess we just need a git url, maybe pointing to something with a _CoqProject file or opam-installable, what do you think @matej-kosik ?

**Comment by JasonGross** [comment]
There might be a part of fiat that builds with 8.6, but I haven't done much with it since the beginning of the summer (but I'll be starting again nowish).  Here's fiat-crypto's _CoqProject: https://github.com/mit-plv/fiat-crypto/blob/master/_CoqProject.  It depends on coqprime, which we bundle: https://github.com/mit-plv/fiat-crypto/blob/master/coqprime/_CoqProject ; it probably works to have either coqprime already installed from OPAM, or to pass `COQPATH="$(pwd)/coqprime"`.

**Comment by herbelin** [comment]
@mattam82: Ah, I now see what you mean. Sorry for not having got it earlier. I was not looking at the right line.

**Comment by herbelin** [comment]
FI, I just tested fiat-crypto and it has no effect.

**Comment by mattam82** [comment]
Cool!
Le mar. 20 sept. 2016 à 14:28, Hugo Herbelin <notifications@github.com> a
écrit :

**Comment by maximedenes** [comment]
This PR seems to have unrelated commits. @herbelin, can you resubmit it with only the relevant ones and base it on v8.6? I'll merge it ASAP. Thanks!

**Comment by ejgallego** [comment]
Yeah, a rebase would be great so we could run Travis on this.

**Comment by Zimmi48** [comment]
@herbelin Are you still planning to work on this? What is the issue that blocked this? Does it need discussion rather than progress?

**Comment by herbelin** [comment]
IIRC, the PR is ok and improve on the regularity of the algorithm for inferring the return clause but it was causing [BZ#5107](https://coq.inria.fr/bugs/show_bug.cgi?id=5107). The latter could itself be resolved by improving the unification algorithm, but I'm somehow waiting on #991 (Unifall) before re-experimenting with unification, and also waiting to know where we are going in terms of improving unification since we are several interested by this question, and where we are going in terms of letting Coq evolve while at the same time maintaining a compatibility.

**Comment by herbelin** [comment]
This PR is a case study about our methods. Do we merge it now knowing the (limited) impact on unification semantics? Do we expect working at the same time on unification to try to have a unification strictly better, without regressions?

**Comment by Zimmi48** [comment]
To my eyes it depends on two questions:
- Is it accurate to consider the change of semantics a "regression"? If the new semantics is better and the scripts are not to complicated to adapt, then we might still push this change.
- If the new behavior is better but it is important to preserve compatibility, is there a way to restore compatibility using an option?

I'm putting a "needs: overlay" label for now supposing the answer is: not an actual regression and scripts can be adapted.

**Comment by JasonGross** [comment]
My impression was that there was no current development impacted (and thus no overlay needed). I would say: note it in CHANGES, and merge.  There's a strictly backwards compatible solution of "include the return clause explicitly", and it's a relatively rare corner case, so I don't think it needs a compatibility option ( though I wouldn't object to one).

**Comment by Zimmi48** [comment]
Oh OK. In any case, a rebase should confirm that no development is impacted.

**Comment by Zimmi48** [comment]
@herbelin Can you rebase?

**Comment by Zimmi48** [comment]
This should also help knowing who the code owners are and who should self-assign this PR.

**Comment by herbelin** [comment]
> Can you rebase?

Was difficult (waited too long w/o a rebase) but finally done.

**Review by mattam82 — APPROVED** [review]
I'll merge when the checks pass

**Comment by SkySkimmer** [comment]
The fiat crypto error looks like a real error. IIUC this PR is intentionally breaking compatibility so we need an overlay.

**Comment by Zimmi48** [comment]
@SkySkimmer AFAICT this error is present in master as well.

**Comment by Zimmi48** [comment]
I may have spoken too fast, I have seen a recent failure of fiat-crypto on master, but a most recent build succeeded.

**Comment by SkySkimmer** [comment]
There was an error on the fiat side which is now fixed, but IIRC it was a different error.
I'm going to rerun the job anyway.

**Comment by Zimmi48** [comment]
> I'm going to rerun the job anyway

I had the same idea, so it's already done ;-)

**Comment by Zimmi48** [comment]
Looks like there is a new error in fiat-crypto. There's also a whitespace error.

**Comment by herbelin** [comment]
> Looks like there is a new error in fiat-crypto.

Indeed. I then reinstalled the overlay and submitted the fix to mit-plv/fiat-crypto#440.

**Comment by herbelin** [comment]
@ppedrot: CI now green, this is ready.

**Pull request merged**
